File: 120-triangle/120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int levels = triangle.size();
        int lastRow = triangle.get(levels-1).size();
        int[][] dp = new int[levels][lastRow + 1];
        for(int[] row : dp) Arrays.fill(row, Integer.MAX_VALUE);
        dp[0][1] = triangle.get(0).get(0);
        
        for(int i = 1;i < levels; i++){
            for(int j = 1; j <= i + 1;j++){
                dp[i][j] = triangle.get(i).get(j-1) + Math.min(dp[i-1][j], dp[i-1][j-1]);
            }
        }
        
        int ans = Integer.MAX_VALUE;
        for(int num : dp[levels-1]) ans = Math.min(ans,num);
        
        return ans;
    }
}

File: 120-triangle/120-triangle.py
class Solution:
    def minimumTotal(self, t: List[List[int]]) -> int:
        @lru_cache
        def helper(i, j):
            if i == len(t)-1:
                return t[i][j]
            
            return t[i][j] + min(helper(i+1, j), helper(i+1, j+1))
                              
        return helper(0, 0)
